docs/design/proposals/encrypted-pvc.md

```diff
@@ -0,0 +1,99 @@
+# Encrypted Persistent Volume Claims
+
+## Proposal
+Subject of this proposal is to add an encrypt feature to ceph-csi in order to allow encryption at rbd volumes level.
+
+Some but not all the benefits of this approach:
+- guarantee encryption in transit to rbd without using messenger v2
+- extra security layer to application with special regulatory needs
+- at rest encryption can be disabled to selectively allow encryption only where required
+
+## Document Terminology
+- volume encryption: encryption of a volume attached by rbd
+- encryption at rest: encryption of physical disk done by ceph
+- LUKS: Linux Unified Key Setup: stores all of the needed setup information for dm-crypt on the disk
+- dm-crypt: linux kernel device-mapper crypto target
+- cryptsetup: the command line tool to interface with dm-crypt
+
+## Proposed Solution
+
+The proposed solution in this document, is to address the volume encryption requirement by using dm-crypt module through cryptsetup cli interface.
+
+### Implementation Summary
+
+- Encryption is implemented using cryptsetup with LUKS extension.
+  A good introduction to LUKS and dm-crypt in general can be found [here](https://wiki.archlinux.org/index.php/Dm-crypt/Device_encryption#Encrypting_devices_with_cryptsetup)
+  Functions to implement necesary interaction are implemented in a separate dmcrypt.go file.
+  1. applyLUKS
+  2. openLUKS
+  3. getPassphrase
+  4. closeLUKS
+
+- `createRBDImage`: refactored to call `applyLUKS`, before returning, if encrypted volume option is set.
+- `applyLUKS`: attaches newly generated image to the worker node where provisioner is running and sets up LUKS using `cryptsetup` command.
+  The function:
+  - resolves the passphrase from a `getPassphrase` function according to kms volume options.
+  - A new passphrase, if not found, is stored in the kms using target volume name (i.e. pvc-8a710f4c934811e9) for lookup
```

Comment:
Will this always be an auto-generated passphrase when requesting a new PVC (each PV w/ its own unique key)? In the case of Vault, that might be fine. For other secret sources, might need a way to store a `StorageClass`-supplied key and/or a `PVC`-supplied key.

Reply:
That's a good question. Would definitely be useful to allow users specify their key id, or refer to it regardless of the source; however, unless there are strong objections, I would probably limit for an initial implementation, to auto-generate passphrases.

Reply:
I'm mostly just concerned about designing a feature that can be cleanly expanded in the future. I think it's fine if the initial implementation only auto-generates passphrases. However, if a `StorageClass` global passphrase is eventually supported, there wouldn't be a way to have a PVC auto-generate its own passphrase unless perhaps a new attribute was added to indicate the passphrase source is "generate" vs "static".

Reply:
are we going to support secret+storageclass based encryption? let's say the user specifies the secret name in the storage class and we are going to fetch the secret details in ceph-csi rbd plugin and use it for encryption.

Reply:
The idea is to have the storageclass define the connection to the kms. Each PVC provisioned with this storage class will then generate a new passphrase and store it / read it from a pre-configured key path. In a second iteration we could allow PVC to specify a key path that overrides the storageclass conguration.
I'm open for any suggestion to improve it anyway.